Implement Tensile Instability Control (TIC) (#769)
trixi-framework/TrixiParticles.jl: docs/src/systems/weakly_compressible_sph.md

Modules = [TrixiParticles]
Pages = [joinpath("schemes", "fluid", "weakly_compressible_sph", "density_diffusion.jl")]
```

## [Particle Shifting Technique](@id shifting)

> The Particle Shifting Technique (PST) is a technique to correct tensile instability

Reviewer: ```suggestion
The Particle Shifting Technique (PST) is a method to correct tensile instability
```

Author: I tried to avoid the word method, as this is what we usually use for the solver itself, so it would be the WCSPH method with the PST and TIC techniques.

Reviewer: Just sounds weird to have technique twice so close repeated.

Author: I asked ChatGPT.

Author: ```suggestion
The Particle Shifting Technique (PST) is used to correct tensile instability.
```

in regions of low pressure, as observed in viscous flow around an object.
Without PST, tensile instability causes non-physical separation of the fluid
from the object, introducing void regions behind the object.

At lower resolutions, PST alone can be effective to correct a viscous flow
around a cylinder, as shown in this figure.
![particle_shifting](https://github.com/user-attachments/assets/70892b0b-af36-4531-b328-f73e63a5e33c)
At higher resolutions, PST alone is not effective anymore; see the figure in

> [Tensile Instability Control](@ref tic).

Reviewer 2: I’m not sure I fully understand the reason for this behavior. Is it related to a smaller compact support at higher resolution?
We also observe detachment from the boundary at coarser resolutions, which creates a free surface that PST cannot handle—correct?
Why does the problem actually get worse as the resolution increases?
Is the same void region also present at coarser resolution, but "compensated" for by the larger compact support?

Author: The initial detachment is about the same size for both the lower and the higher resolution simulation. But for the lower resolution, the void region is much smaller compared to the particle size, so the shifting can close the hole again, whereas in the higher resolution it is too large to close. 

Another way to think about it is that in the high resolution simulation we have more particles where the velocity is essentially orthogonal to the boundary.

> We generally recommend using PST and Tensile Instability Control together
> in such a simulation.

Reviewer: ```suggestion
We recommend using PST and Tensile Instability Control together
in such simulations.
```

### Mathematical formulation

We use the following formulation by [Sun et al. (2018)](@cite Sun2018).
After each time step, a correction term ``\delta r_a`` is added to the position ``r_a``
of particle ``a``, which is given by
```math
\delta r_a = -4 \Delta t \, v_\text{max} h
    \sum_b \left( 1 + R \left( \frac{W_{ab}}{W(\Delta x_a)} \right)^n \right) \nabla W_{ab}
    \frac{m_b}{\rho_a + \rho_b},
```
where:
- ``\Delta t`` is the time step,
- ``v_\text{max}`` is the maximum velocity over all particles,
- ``h`` is the smoothing length,
- ``R`` and ``n`` are constants, which are set to ``0.2`` and ``4`` respectively,
- ``W(\Delta x_a)`` is the smoothing kernel of the particle size of particle ``a``,
  which can be interpreted as the target particle spacing that we want to achieve.
- ``\nabla W_{ab}`` is the gradient of the smoothing kernel,
- ``m_b`` is the mass of particle ``b``,
- ``\rho_a, \rho_b`` is the density of particles ``a`` and ``b``, respectively.

Note that we replaced ``\text{CFL} \cdot \text{Ma}`` by ``\Delta t \cdot v_\text{max} / h``,

> as explained on page 29, right above Equation 9.

Reviewer 2: ```suggestion
as explained in [Sun2018](@cite Sun2018) on page 29, right above Equation 9.
```
